storage/range.go

```diff
@@ -1425,62 +1519,55 @@ func (r *Range) InternalTruncateLog(batch engine.Engine, ms *proto.MVCCStats, ar
 	reply.SetGoError(err)
 }
 
-// InternalLeaderLease evaluates and responds to a request to grant a leader lease.
-func (r *Range) InternalLeaderLease(args *proto.InternalLeaderLeaseRequest, reply *proto.InternalLeaderLeaseResponse) {
-	// TODO(tschottdorf) stub for now to get tests working.
-	r.setLease(&args.Lease)
-	// r.grantLeaderLease(args.Lease)
-}
+// InternalLeaderLease sets the leader lease for this range. After a
+// lease has been set, calls to HasLeaderLease() will return true if
+// this replica is the lease holder and the lease has not yet
+// expired. If this range replica is already the lease holder, the
+// expiration will be extended as indicated. Otherwise, all duties
+// required of the range leader are commenced, including gossiping
+// cluster and config info where appropriate and clearing the command
+// queue and timestamp cache.
+func (r *Range) InternalLeaderLease(batch engine.Engine, ms *proto.MVCCStats, args *proto.InternalLeaderLeaseRequest, reply *proto.InternalLeaderLeaseResponse) {
+	r.Lock()
+	defer r.Unlock()
 
-// requestLeaderLease sends a request to obtain or extend a leader lease for
-// this replica. Being a first mover, it registers itself as a task with the
-// stopper.
-func (r *Range) requestLeaderLease(term uint64) {
-	if !r.stopper.StartTask() {
+	// Verify details of new lease request. The start of this lease must
+	// be afer the prior lease's expiration.
+	oldLease := r.getLease()
+	if !args.Lease.Start.Less(args.Lease.Expiration) {
+		reply.SetGoError(util.Errorf("lease %s - %s invalid", args.Lease.Start, args.Lease.Expiration))
+		return
+	} else if oldLease.RaftNodeID == args.Lease.RaftNodeID {
+		if !oldLease.Start.Less(args.Lease.Start) {
```

Comment:
Actually, I guess there's no harm in a lease holder reducing the time remaining on its lease, and this will be necessary for orderly shutdown without waiting for all leases to expire. We should add a comment (and test) for this scenario.